docs/src/transactions/transaction-kernel.md

```diff
@@ -0,0 +1,104 @@
+# The Transaction Kernel Program
+The transaction kernel program is responsible to execute a Miden rollup transaction within the Miden VM. Therefore, it is written in [MASM](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/main.html) and it is defined as MASM [kernel](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/execution_contexts.html#kernels). The kernel provides context-sensitive security preventing unwanted read and write access. It defines a set of procedures which can be invoked from other [contexts](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/execution_contexts.html#execution-contexts) - e.g., notes - to be executed in the root context.  
+
+In general, the kernel's procedures must reflect everything users might want to do in executing transactions, from transferring assets to complex smart contract interactions with custom code. Learn more about available procedures and contexts [here](transaction-procedures.md).
+
+The kernel has a well-defined structure which must do the following:
+
+1. **Prologue**: prepares the transaction for processing by parsing the transaction data and setting up the root context.
+2. **Note Processing**: executes the note processing loop which consumes each InputNote and invokes the note script of each note.
+3. **Transaction Script Processing**: executes the optional transaction script.
+4. **Epilogue**: finalizes the transaction by computing the created notes commitment, the final account hash, asserting asset invariant conditions and asserting the nonce rules are upheld.
+
+<p align="center">
+    <img src="../diagrams/architecture/transaction/Transaction_program.png" style="width: 75%;">
+</p>
+
+## Memory Layout 
+Miden Assembly doesn't provide dynamic memory allocation. In total, a MASM program has `2^{32}` memory slots. Therefore, we need to pre-allocate areas in which we store all necessary information. The Miden kernel uses memory offsets to manage the allocation. The root context, in other words the kernel, has access to all of those memory slots. 
```

Comment:
I think a big part of the info in the paragraph is probably extraneous. For example, the fact that we lay out data in memory is not related to dynamic memory allocation (which can work the same way as on other architectures). Also, not sure we need to mention total size of memory.

The main reason we write data to memory is so that we can read it later (during note execution, epilogue etc.).

Reply:
I removed the paragraph and put the diagram below